marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,734 @@
+        $ counterexample (show utxos)
+        $ either (const theProperty) (flip counterexample theProperty . show) result
+
+violations :: MarloweVersion v -> MarloweContext v -> Chain.UTxOs -> TxConstraints v -> TxBodyContent BuildTx BabbageEra -> [String]
+violations marloweVersion marloweContext utxos constraints txBodyContent = fold
+  [ ("mustMintRoleToken: " <>) <$> mustMintRoleTokenViolations marloweVersion constraints txBodyContent
+  , ("mustSpendRoleToken: " <>) <$> mustSpendRoleTokenViolations marloweVersion utxos constraints txBodyContent
+  , ("mustPayToAddress: " <>) <$> mustPayToAddressViolations marloweVersion constraints txBodyContent
+  , ("mustSendMarloweOutput: " <>) <$> mustSendMarloweOutputViolations marloweVersion marloweContext constraints txBodyContent
+  , ("mustSendMerkleizedContinuationOutput: " <>) <$> mustSendMerkleizedContinuationOutputViolations marloweVersion constraints txBodyContent
+  , ("mustPayToRole: " <>) <$> mustPayToRoleViolations marloweVersion marloweContext constraints txBodyContent
+  , ("mustConsumeMarloweOutput: " <>) <$> mustConsumeMarloweOutputViolations marloweVersion marloweContext constraints txBodyContent
+  , ("mustConsumePayouts: " <>) <$> mustConsumePayoutsViolations marloweVersion marloweContext constraints txBodyContent
+  , ("requiresSignature: " <>) <$> requiresSignatureViolations marloweVersion utxos constraints txBodyContent
+  , ("requiresMetadata: " <>) <$> requiresMetadataViolations marloweVersion constraints txBodyContent
+  ]
+
+check :: Alternative m => Bool -> a -> m a
+check condition msg = msg <$ guard (not condition)
+
+mustMintRoleTokenViolations
+  :: MarloweVersion v -> TxConstraints v -> TxBodyContent BuildTx BabbageEra -> [String]
+mustMintRoleTokenViolations MarloweV1 TxConstraints{..} TxBodyContent{..} = fold
+  [ mintsOneToken
+  , sendsOneTokenAndOnlyToken
+  , consumesUtxo
+  ]
+  where
+    consumesUtxo = case roleTokenConstraints of
+      MintRoleTokens txOutRef _ _ -> do
+        check
+          (any ((== txOutRef) . fromCardanoTxIn . fst) txIns)
+          ("UTxO not consumed: " <> show (Chain.renderTxOutRef txOutRef))
+      _ -> []
+
+    sendsOneTokenAndOnlyToken = case roleTokenConstraints of
+      MintRoleTokens _ _ distribution -> do
+        (assetId, address) <- Map.toList distribution
+        (("roleToken: " <> show assetId <> ": ") <>) <$> do
+          let
+            cardanoAssetId = toCardanoAssetId assetId
+            matches (TxOut outAddress (TxOutValue MultiAssetInBabbageEra value) _ _)
+              | selectAsset value cardanoAssetId > 0 = Just (outAddress, value)
+              | otherwise = Nothing
+            matches (TxOut _ (TxOutAdaOnly era _) _ _) = case era of
+          let matchingOuts = mapMaybe matches txOuts
+          case matchingOuts of
+            [(outAddress, value)] -> do
+              fold
+                [ check
+                    (fmap fst (valueToList value) == [cardanoAssetId])
```

Comment:
Is this really needed? Why can't the output contain irrelevant tokens? The coin-selection and balancing algorithms could (in principle) decide to consolidate outputs to the same address into a single output. Irrelevant tokens might be in that output if they had to be consumed in input. In practice, the coin selection and balancing algorithms don't do this, but a future version might.

Reply:
This is checking the output of the constraint solver only (before coin selection and balancing). It's not strictly needed, but it does enforce that the outputs are created cleanly (the constraint solver could decide to aggregate payouts and role tokens, for example, but having those separate makes consuming them easier).

I'm certainly open to a debate on this point however.